### ISSUE TITLE
Multi-level redundancies for all the "apps" for users, like rtorrent

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#14]
rtorrent watchdogs are multiple levels. We should do something similar for deluge + qbittorrent + any docker containers users may run.
It has proven reliable and robust over the years, a software which is known to be flaky, crashy, unreliable rarely gathers any tickets for the past 8 years or so about issues.

So refactoring and reusing that code, similar to the "app launchers" in www/ directory in issue #10 

It's not common for deluge + qbittorrent to crash, but additional redundancy is always good.

Need to be careful not to mess rTorrent tho, since it's flakyness necessitated this multiple levels.

Also checks ought to be faster than they are currently, i believe 2nd level might have upto 15minute delay right now.

**Comment by MagnaCapax** [comment]
## Adversarial Review (Update)

**Issue:** #14 — Multi-level redundancies for all the "apps" for users, like rtorrent
**Reviewed:** 2026-04-19
**Previous review:** 2026-04-04 (NEEDS-INVESTIGATION — see prior comment for full architecture map)
**Verdict:** NEEDS-INVESTIGATION

This update adds materially new evidence from post-review activity. The prior review's architecture map and gap analysis remain accurate; read it first.

---

### New Evidence Since April 4

**Zombie-state failure mode confirmed in production.** Issue #431 ("checkRtorrent: false restart on SCGI timeout during cgroup memory pressure") was filed from a customer support ticket and labeled `complete-verify` after fix. The incident: under cgroup v1 memory pressure, the kernel aggressively reclaims page cache pages, stalling all processes in the user slice. During reclaim stalls, rtorrent's SCGI socket becomes temporarily unresponsive. The pre-fix watchdog interpreted SCGI timeout as "process is dead" and restarted it. The process was alive. The restart caused data corruption on active large torrents (SIGTERM + wait → `rtorrentProcessRestart` tears down an active session mid-write).

This is the zombie-state failure mode the prior review flagged as "theoretical." It is not theoretical. It is a confirmed production incident triggered by the same cgroup pressure conditions that affect all user-space processes on these hosts.

**Implementation pattern established by `3292e954`.** The fix for #431 adds a process-liveness gate to `checkRtorrent.php` at the SCGI timeout branch (around line 474):

```
SCGI timeout threshold reached
  → pgrep: is rtorrent process still alive?
      YES → refresh grace timestamp, log, continue (do NOT restart)
      NO  → cleanup stale socket, then restart
```

This pattern — API/socket health check, then process-liveness confirmation before acting — is now the established template for any watchdog that adds an API layer on top of a process-existence check.

---

### Updated Adversarial Findings

**[CORRECTNESS — C] Any API health check for deluged/qbittorrent-nox MUST include the process-liveness gate.** The #431 incident proves that acting on an API timeout without confirming process death causes data corruption. The deluged RPC API and qbittorrent-nox WebUI API will experience the same transient unresponsiveness under memory pressure (they are subject to the same cgroup constraints). An implementation that restarts on API timeout without a `pgrep` confirmation pass repeats the #431 bug. This constraint was not in the issue body.

**[CORRECTNESS — M] The current deluge watchdog uses `killall -9` (SIGKILL) by default.** `pmssUserWatchdogRestartProcessesIf` defaults `$signal = 9`. For rtorrent, `rtorrentProcessRestart` sends SIGTERM first, waits, then SIGKILL. For deluged, the watchdog calls `killall -9` directly (no graceful-stop phase). A deluged SIGKILL during an active write is at least as disruptive as the rtorrent false restart. If a health check layer causes deluged to be killed more frequently (including false positives), this amplifies the blast radius.

**[SCOPE — M] Docker container monitoring: still unfeasible.** No container manifest mechanism has been added since the April 4 review. The feasible scope remains: "daemon is running per enabled user." Individual container monitoring is a separate design problem requiring a container registry of intended state. This remains a candidate for a separate issue.

---

### Revised Big Questions

1. **Worst case if implemented without the liveness gate**: API health check restarts deluged on every cgroup memory-reclaim stall. Active transfers are torn down by SIGKILL. Users observe data corruption. The prior rtorrent pattern (120-180s grace + process liveness check) exists precisely to prevent this — omitting it from a new implementation is not a simplification, it is the bug.

2. **Inversion — maximum damage implementation**: Add a deluged RPC health check with a 5-second timeout and immediate SIGKILL on timeout, no process liveness check, no grace period. Users under memory pressure would experience data corruption on every memory reclaim cycle. This is exactly what the #431 fix was written to prevent for rtorrent.

3. **Regret in 6 months**: Adding API health checks without grace periods and process-liveness gates (i.e., copy-pasting the pre-#431 rtorrent watchdog logic) creates a new class of operator-reported corruption incidents that are harder to diagnose because "the watchdog is working correctly" per its own logs.

4. **Removal inversion**: If a "deluged API health check with process-liveness gate" already existed, we would absolutely keep it — #431 proves this failure mode is real and the mitigation is necessary.

5. **Requirements challenge**: The issue says "not common for deluge/qbittorrent to crash." This is correct — but the failure mode #431 exposes is not "app crash," it is "app zombie under load." The requirements should be reframed around zombie detection, not crash recovery. The current pgrep-only watchdog handles crashes adequately.

6. **Deletion opportunity**: The per-user cron layer for rtorrent (`.rtorrentRestart.php`) adds marginal value over the root cron layer, which already runs every 2 minutes. Adding a per-user layer for deluge/qbittorrent at the same interval as root cron doubles check frequency without halving time-to-recovery in any meaningful way. Omit the per-user layer; invest in the API health check + liveness gate at root cron level instead.

---

### Investigation Results (Updated)

**Affected files:**
- `scripts/cron/checkRtorrent.php` (511 lines, post-fix) — reference implementation, process-liveness gate at line 474
- `scripts/cron/checkDelugeInstances.php` — unchanged since prior review; pgrep only
- `scripts/cron/checkQbittorrentInstances.php` — unchanged since prior review; pgrep only
- `scripts/lib/userLifecycle.php:241` — `pmssUserWatchdogRestartProcessesIf`: kills with signal 9 by default
- `scripts/lib/rtorrent/process.php` (558 lines) — state tracking library; now 54 lines longer than prior review
- `scripts/lib/rtorrent/scgi.php:303` — `rtorrentScgiPing()` — API health check reference

**New git context:**
- `3292e954` (2026-04-16) — process-liveness gate for SCGI timeout path; contract test added in `checkRtorrentStaleSocketContractTest.php`
- `693df1a9` (2026-04-16) — Deluge managed config refresh; unrelated to watchdog behavior
- Deluge/qbit watchdog files unchanged since prior review

**Tests:** `scripts/lib/tests/development/checkRtorrentStaleSocketContractTest.php` covers the new liveness-gate behavior. No equivalent tests exist for deluge or qbittorrent watchdog paths.

---

### Recommendations

The investigation is now more complete. Implementation can proceed for the non-docker scope if the issue is updated to:

1. **Replace the vague "multiple levels" framing** with the concrete gap: "add API health check + process-liveness gate + grace period to deluged and qbittorrent-nox watchdogs, matching the pattern established in `checkRtorrent.php:474` as of commit `3292e954`."

2. **Specify deluged RPC credential handling** — deluged RPC requires authentication; the watchdog connection needs credentials without exposing them in process args or log output.

3. **Split the docker ask** into a separate issue — it requires a container manifest design that does not exist yet.

4. **Signal escalation** — consider whether SIGTERM-before-SIGKILL should be applied to deluged restarts (consistent with qbittorrent-nox, which already uses a custom terminator in `checkQbittorrentInstances.php`).

---

### Worth-Implementing Analysis

- **Cost**: Medium-high. Proper implementation requires: deluged RPC client (with auth), qbittorrent-nox WebUI API client (HTTP), grace period machinery (state files per user), process-liveness gate, and failure counters. Test coverage for each. Minimum 300-500 LOC new code across library modules.
- **Benefit**: Confirmed by production incident (#431). The zombie failure mode under cgroup memory pressure is real. Users experience data corruption from false restarts — having the health check without the liveness gate is worse than no health check at all.
- **Alternative**: The immediate safety improvement is ensuring deluged restarts use SIGTERM-before-SIGKILL (matching qbittorrent's existing custom terminator). This is 5 lines, zero API client complexity, and reduces the blast radius of the existing pgrep-triggered restarts. File as a separate small issue.
- **Verdict rationale**: The API health check + liveness gate work is justified by confirmed production incidents. The complexity is real but bounded. The docker scope must be removed before implementation starts.

---

### Anti-Pattern Hunt

- `REGEX_FOR_SEMANTICS`: **PASS** — no regex-for-semantics proposed
- `PREMATURE_OPTIMIZATION`: **PASS** — addresses confirmed failure mode, not speculation
- `OPERATOR_HANDOFF_ANTIPATTERN`: **PASS** — no handoff queue / alert theater
- `FALSE_COMPLETION_CLAIM`: **PASS** — no completion claims
- `FABRICATED_AUTHORITY_OR_PATH`: **PASS** — all cited paths verified in source
- `FAIL_OPEN_SAFETY_BEHAVIOR`: **MILD CONCERN** — the *absence* of a process-liveness gate in a future implementation would be fail-open (restart deluged on any API timeout, including false positives). The implementation MUST include the liveness gate. An implementation without it is a severity-C safety regression.
- `DRY_RUN_THEATER`: **PASS** — not present
- `INVESTIGATION_LOOP_THEATER`: **MILD CONCERN** — this is the second NEEDS-INVESTIGATION verdict. Avoid a third. The issue should be decomposed: (a) file a focused issue for deluged/qbittorrent API health check + liveness gate, (b) close or update the docker container scope, (c) close this meta-issue once sub-issues are filed.

---

### Decomposition Targets (Updated)

- **[REFACTOR]** `scripts/lib/rtorrent/process.php` (558 lines) — further grown since prior review; state-tracking machinery could become `scripts/lib/watchdog/state.php` for reuse across all app watchdogs. Above AGENTS.md LOC guideline.
- **[REFACTOR]** `scripts/cron/checkRtorrent.php` (511 lines) — still above threshold. Decomposition would enable the process-liveness gate pattern to be extracted and reused by deluge/qbittorrent health checks without duplication.
- **[NEW-ISSUE CANDIDATE]** `scripts/lib/userLifecycle.php:241` — deluged restarts use `killall -9` with no SIGTERM grace. A 5-line fix (add `$signal = 15` + terminator callable for deluged) reduces data corruption risk from existing pgrep-triggered restarts without requiring the full API health check implementation.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review (Third Pass)

**Issue:** #14 — Multi-level redundancies for all the "apps" for users, like rtorrent
**Reviewed:** 2026-04-26
**Previous reviews:** 2026-04-04 (NEEDS-INVESTIGATION), 2026-04-19 (NEEDS-INVESTIGATION)
**Verdict:** REWORK

---

### Why a Third Review

The April 19 review explicitly warned against a third NEEDS-INVESTIGATION verdict and called for issue decomposition. In the seven days since, no sub-issues were filed, the issue spec was not updated, and the key watchdog files remain unchanged. Issuing a third NEEDS-INVESTIGATION without new action would be the investigation-loop anti-pattern — the prior two reviews have fully mapped the problem; what is missing is a concrete, implementable issue spec that Codex can act on. This review breaks the loop by (a) escalating to REWORK, (b) filing the two concrete sub-issues the April 19 review identified, and (c) supplying the implementation spec with patterns established since April 19.

---

### New Evidence Since April 19

**`scripts/lib/lighttpd/watchdogSocketProbe.php` (commit `8e766a9f`, 2026-04-24, Refs #469)** establishes the canonical PMSS pattern for connection-based health checks:

```php
// Retry budget: 4 attempts, 2s delay between retries, 5s per-connect timeout.
// Injectable $probe and $sleep callables enable unit testing without real sockets.
function pmssLighttpdWatchdogSocketProbeWithRetry(string $socketPath, array $options = array()): array
```

The structure — configurable attempt count, sleep between retries, injectable test doubles — is the template any deluged RPC or qbittorrent-nox WebUI health check must follow. This file did not exist in the April 19 investigation.

**deluged still uses SIGKILL with no terminator.** The April 19 review identified this as an immediate 5-line fix. Confirmed in current source:

```php
// checkDelugeInstances.php — signal defaults to 9, no terminator argument
pmssUserWatchdogRestartProcessesIf($thisUser, ..., 'deluge restarted to apply upload throttle');
//                                                 ^^^^ no $signal, no $terminator
```

qbittorrent-nox already uses signal=15 + a custom SIGTERM terminator (`killall -u $user -TERM qbittorrent-nox`). Deluged does not. This asymmetry predates this issue and creates an unnecessary data-corruption risk on every pgrep-triggered restart today — no API health check required to fix it.

---

### Adversarial Findings

**[INVESTIGATION_LOOP — C] Two NEEDS-INVESTIGATION verdicts with no follow-through is the anti-pattern.** The April 19 review's recommendations (decompose into sub-issues, narrow docker scope, file SIGTERM fix) were not acted on. Continuing to label this NEEDS-INVESTIGATION without taking the decomposition action is process theater. The investigation is complete. The blocker is a stale issue spec, not missing knowledge.

**[REQUIREMENTS — C] The issue spec has not been updated since filing in 2024.** It references the state of watchdogs that have since changed materially: the 15-minute delay concern is resolved, basic process watchdogs for deluge and qbittorrent already exist, and the zombie-state failure mode (previously theoretical) was confirmed by incident #431 (cgroup memory pressure, false SCGI restart). The spec still reads as if none of this exists.

**[SCOPE — C] Docker container monitoring remains architecturally unfeasible as scoped.** This has been stated in both prior reviews. No container manifest mechanism has been introduced. The feasible scope — daemon running per enabled user — is already implemented in `checkRootlessDocker.php`. The per-container ask requires a separate design and a separate issue. Keeping it in this issue prevents the issue from ever reaching PROCEED.

**[CORRECTNESS — M] The immediate safety gap (deluged SIGKILL) is smaller and cheaper than the issue implies.** The issue frames this as "multi-level redundancy" work. The most urgent gap — that a pgrep-triggered restart today kills deluged with SIGKILL, potentially mid-write — is a 3-line fix that doesn't require any API health check infrastructure. Conflating it with the larger API health check work delays the quick win.

**[ENGINEERING — M] Grace period handling for deluged and qbittorrent-nox must mirror the rtorrent pattern.** Commit `3292e954` established the contract: API/socket timeout → pgrep confirmation → if process alive, refresh grace timestamp and continue; only restart on confirmed process death. Any deluged RPC probe or qbittorrent WebUI probe that acts on a timeout without the liveness check repeats the pre-`3292e954` bug. The `pmssLighttpdWatchdogSocketProbeWithRetry` pattern handles the retry budget; the process-liveness gate (from `checkRtorrent.php:474–495`) must be layered on top.

---

### Big Questions (Abbreviated — Full Analysis in April 4 and April 19 Comments)

1. **Worst case without the liveness gate**: Deluged enters a transient API-unresponsive state under cgroup memory pressure → health check sees timeout → SIGKILL → active 100GB transfer torn down mid-write. This is identical to the #431 incident. Confirmed failure mode, confirmed blast radius.

2. **Worst case of the SIGKILL gap today** (no health check needed to trigger): A pgrep watchdog catches deluged during its normal restart cycle after a crash, confirms it is not running, starts it — this is the safe path. But if deluged is slow to exit and `pgrep` catches it while still alive, SIGKILL interrupts a still-running write. Low probability but non-zero, and the fix is trivial.

3. **Regret in 6 months**: Implementing a deluged RPC health check without injectable test doubles (mirroring `watchdogSocketProbe.php`) results in a watchdog that cannot be unit-tested, like the pre-2026 rtorrent watchdog. The `watchdogSocketProbe.php` pattern exists precisely to prevent this — any new API health check must use the same injectable structure.

4. **Deletion opportunity**: The per-user cron layer for deluge/qbittorrent was questioned in April 19 review and remains questionable. Root cron runs every 1 minute for deluge. Adding a per-user layer at the same interval doubles check frequency without meaningfully shortening time-to-recovery. Omit the per-user layer and focus on the API health check at root cron level.

---

### Implementation Spec (for Codex, once issue is reworked)

This review files two sub-issues with the concrete scopes. For each:

**Sub-issue A — deluged SIGTERM grace (trivial, file first):**
- `checkDelugeInstances.php`: add `signal=15` parameter and a SIGTERM terminator to the `pmssUserWatchdogRestartProcessesIf` call, mirroring qbittorrent's existing pattern. Template from `checkQbittorrentInstances.php`:
  ```php
  // qbittorrent already has this:
  static function () use ($thisUser): void {
      passthru('killall -u '.escapeshellarg($thisUser).' -TERM qbittorrent-nox 2>/dev/null');
  }
  ```
  Adapt for deluged and deluge-web (two processes; terminate deluged first, then deluge-web). No new library required.

**Sub-issue B — API health check + liveness gate (medium, file second):**
- Pattern: `scripts/lib/lighttpd/watchdogSocketProbe.php` → adapt as `scripts/lib/deluge/rpcProbe.php` and `scripts/lib/qbittorrent/apiProbe.php`. Use same injectable structure (probe callable, sleep callable).
- Layer on top: process-liveness gate from `checkRtorrent.php:474–495`. If API probe fails but pgrep still finds the process: refresh grace timestamp, log, continue. Only restart on confirmed process death.
- Deluged RPC requires authentication credentials. Do not pass them via process arguments (ADR 0004: shell command guardrails apply). Credential handling approach must be specified in the sub-issue.
- Add contract tests analogous to `checkRtorrentStaleSocketContractTest.php`.

---

### Worth-Implementing Analysis

- **SIGTERM fix**: Cost = 3 lines + test. Benefit = eliminates SIGKILL on active deluged sessions. Verdict: PROCEED on sub-issue A the moment it is filed.
- **API health check**: Cost = 300–500 LOC including library modules and tests. Benefit = confirmed by #431 — zombie-state under cgroup pressure is real, not theoretical. Verdict: PROCEED on sub-issue B, but only after sub-issue A is merged and the docker scope is removed from this meta-issue.
- **This meta-issue**: After sub-issues A and B are filed, this issue's remaining value is coordination tracking. Consider closing it and referencing the sub-issues once filed.

---

### Anti-Pattern Hunt

- `REGEX_FOR_SEMANTICS`: **PASS** — no regex-for-semantics in this proposal
- `PREMATURE_OPTIMIZATION`: **PASS** — zombie-state confirmed by production incident
- `OPERATOR_HANDOFF_ANTIPATTERN`: **PASS** — no alert queue theater
- `FALSE_COMPLETION_CLAIM`: **PASS** — no completion claims in issue history
- `FABRICATED_AUTHORITY_OR_PATH`: **PASS** — all file paths verified in current source
- `FAIL_OPEN_SAFETY_BEHAVIOR`: **CONCERN — see findings** — any health-check implementation without a process-liveness gate defaults to restarting on API timeout (fail-open on false positives). This is the #431 bug reintroduced. The liveness gate is not optional.
- `DRY_RUN_THEATER`: **PASS** — not present
- `INVESTIGATION_LOOP_THEATER`: **FAIL — this review breaks the loop** — two prior NEEDS-INVESTIGATION verdicts with no sub-issue filing or spec update. This review files the sub-issues and escalates to REWORK.

---

### Caller Analysis

Not applicable — this issue proposes additions, not removals.

---

### Decomposition Targets

- **[REFACTOR]** `scripts/lib/rtorrent/process.php` (558 lines) — state-tracking machinery (grace periods, failure counters, escalation, state files) is rtorrent-specific but could become `scripts/lib/watchdog/state.php`, reusable across all app watchdogs. Above AGENTS.md LOC split guideline (75 lines).
- **[REFACTOR]** `scripts/cron/checkRtorrent.php` (511 lines) — SCGI health check, grace period logic, process-liveness gate, and restart orchestration are distinct responsibilities. Extracting the liveness gate into a shared library enables sub-issue B without copy-pasting from a 511-line file.

---

### Recommendations

Before any implementation:

1. **File sub-issue A** (SIGTERM grace for deluged) — trivial, ship immediately.
2. **File sub-issue B** (API health check + liveness gate) — medium, needs RPC credential design.
3. **Update or close this meta-issue** — remove docker container scope, or move it to a separate issue with a container manifest design. The meta-issue should not remain open indefinitely waiting for docker work that has no design.
4. **Do not file a fourth adversarial review on this issue in its current form.** The analysis is complete. The next step is action, not more investigation.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review (Fourth Pass)

**Issue:** #14 — Multi-level redundancies for all the "apps" for users, like rtorrent
**Reviewed:** 2026-05-29
**Previous reviews:** 2026-04-04 (NEEDS-INVESTIGATION), 2026-04-19 (NEEDS-INVESTIGATION), 2026-04-26 (REWORK)
**Verdict:** SHOULD-NOT-IMPLEMENT

This pass is brief by design. The prior three reviews are the investigative record; this pass resolves the open loop.

---

### Why a Fourth Review

The April 26 review said "do not file a fourth review on this issue in its current form" — meaning: break the investigation loop, not "never review again." The sub-issues were filed that same day, satisfying the REWORK requirement. Thirty-three days have passed with no operator action: the issue remains open, the label is stale, and a material code change (May 29 refactor) post-dates the last review. This pass closes the loop.

---

### New Evidence Since April 26

**Sub-issues confirmed filed on 2026-04-26 (same day as the third review):**
- **#477** — deluged watchdog: replace SIGKILL with SIGTERM grace before hard kill. OPEN. PROCEED verdict since 2026-04-27 (reconfirmed 2026-05-19, 2026-05-26). **Not yet implemented** — 7+ failed build attempts (predominantly `context_exhausted`).
- **#478** — deluged/qbittorrent-nox: add API health-check + process-liveness gate to watchdogs. OPEN. REWORK verdict — 5 spec items remain unresolved (probe type, credential source, grace state files, deluge-web scope, zombie kill signal for qbittorrent-nox).

**May 29 refactor commit `a577667d`** restructured `userLifecycle.php` into focused modules (`user/watchdog.php`, `user/identity.php`, `user/selection.php`). `checkDelugeInstances.php` was updated to include `user/watchdog.php` instead of `userLifecycle.php`. The signal call on line 22 is unchanged. This commit post-dates the May 26 verification on #477 — see the #477 build-unblock note posted alongside this review.

**Docker container scope has not been filed as a separate issue.** The April 26 review explicitly recommended doing so. This gap remains.

---

### Adversarial Findings

**[LABEL_STALENESS — C] Issue carries `needs-investigation` but the April 26 verdict was REWORK.** The label was never updated. The Build agent's decision logic skips `should-not-implement`-labeled issues and routes `needs-investigation` issues to investigation cycles. This meta-issue in `needs-investigation` state invites a fifth investigation cycle that adds no value.

**[SCOPE_INCOMPLETION — M] Docker container monitoring scope was never closed out.** Three reviews flagged it as architecturally unfeasible as stated (no container manifest, no ground-truth expected state). The April 26 review said to move it to a separate issue with a container manifest design. It was not filed. The meta-issue's open state partially stems from this unresolved scope item — there is no implementation path for this third part of the original request.

**[BUILD_FRICTION — M] Sub-issue #477 has 7 consecutive build failures for a 2-line change.** The fix is a unique-substring replacement in one 24-line file plus a test array update in a 101-line file. Context exhaustion in a sandboxed build session for this scope is anomalous. The May 29 commit (`a577667d`) updated `checkDelugeInstances.php` after the last verification — a build agent noticing this post-dates the May 26 review may hesitate. Verification: the commit changed only the `require_once` path (line 13); the signal call on line 22 is unchanged and the unique substring appears exactly once. See the #477 build-unblock note for the updated implementation spec.

---

### Big Questions (Abbreviated — Full Treatment in Prior Reviews)

1. **Worst case if #14 implemented directly:** An agent reads the 2024 issue body without reading the three review comments and re-implements work that already exists — creating duplicate logic or conflicting with the in-progress #477/#478 work. The meta-issue spec is now a trap, not a guide.
2. **Deletion opportunity:** Close this issue. No code needed.
3. **Requirements challenge:** The issue requirements were written in 2024 before basic deluge and qbittorrent process watchdogs existed. The requirements are stale by design — the world changed around them.
4. **Regret test:** In 6 months, an open meta-issue whose sub-issues are merged and closed will be confusing noise. Close it now.

---

### Caller Analysis

Not applicable — this is a meta-issue proposing additions, not removals.

---

### Worth-Implementing Analysis

- **This meta-issue as scoped:** SHOULD-NOT-IMPLEMENT. The decomposition is correct; the sub-issues are the implementation path. No agent should implement #14 directly.
- **#477 (SIGTERM grace):** Confirmed worth implementing. 2-line fix. Trivial cost, concrete benefit (eliminates SIGKILL on active deluged writes).
- **#478 (API health check):** Worth implementing once spec items are resolved. Confirmed production failure mode.
- **Docker scope:** Cannot be implemented as stated. Requires a separate design issue with a container manifest mechanism.

---

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS:** PASS — not present.
- **PREMATURE_OPTIMIZATION:** PASS — zombie failure mode confirmed by production incident (#431).
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — sub-issues are self-contained.
- **FALSE_COMPLETION_CLAIM:** PASS — no completion claims in issue history.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — all file paths verified from current source.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS on meta-issue; flagged separately on #478 (probe timeout must not default to restart).
- **DRY_RUN_THEATER:** PASS.
- **INVESTIGATION_LOOP_THEATER:** **RESOLVED** — this pass closes the loop by escalating to SHOULD-NOT-IMPLEMENT and recommending closure. Four NEEDS-INVESTIGATION/REWORK cycles on a meta-issue is the anti-pattern; the cure is not a fifth review but a verdict that removes the issue from the build candidate pool.
- **LACK_OF_INVESTIGATION:** PASS — counter-signals checked: #477 current state (12 comments, PROCEED), #478 current state (5 comments, REWORK), codebase file state post-May-29 commit.

---

### Decomposition Targets

Prior reviews identified these targets; unchanged:

- **[REFACTOR]** `scripts/cron/checkRtorrent.php` (~511 lines) — grace period logic, SCGI health check, and process-liveness gate are separable responsibilities. Extraction is warranted when #478 lands (third reuser of the pattern).
- **[REFACTOR]** `scripts/lib/rtorrent/process.php` (~558 lines) — watchdog state machinery could become a shared `scripts/lib/watchdog/state.php`. Blocked on #478.
- **[REFACTOR]** `scripts/lib/lighttpd/watchdogSocketProbe.php` — injectable retry probe pattern is a candidate for a generic probe library when #478 adds two more reusers.

---

### Recommendations

1. **Close this issue.** The work is correctly tracked in #477 (SIGTERM grace) and #478 (API health check). A meta-issue whose implementation path lives entirely in sub-issues adds no tracking value while remaining open.
2. **File the Docker container monitoring scope as a new issue** if it remains desired — with a container manifest design as prerequisite (ground-truth expected state). The title should reflect the design constraint: "Docker container watchdog: requires container manifest for per-user expected state."
3. **See the #477 build-unblock note** for the updated implementation spec reflecting the May 29 file state.

— Väinämöinen
